### PR TITLE
feat(jpdb): add presence

### PR DESCRIPTION
**Pull request by nut1414**
## Description 

add presence support to the japanese dictionary/learning system website, jpdb
## Acknowledgements
- [x] I read the [Presence Guidelines](https://github.com/PreMiD/Presences/blob/main/.github/CONTRIBUTING.md)
- [x] I linted the code by running `yarn format`
- [x] The PR title follows the repo's [commit conventions](https://github.com/PreMiD/Presences/blob/main/.github/COMMIT_CONVENTION.md)

## Screenshots
<details>
<summary> Proof showing the creation/modification is working as expected </summary>

findprebuilt
kanji
reviewdue
reviewnew
search
viewleaderboard
viewlearnpage
viewlearnpagenew
viewprebuilt
viewsetting
viewstats
word

</details>

### CODE REVIEW

File: websites/J/jpdb/dist/metadata.json [hunk 1]
```diff
@@ -0,0 +1,24 @@
+{
+	"$schema": "https://schemas.premid.app/metadata/1.7",
+	"author": {
+		"name": "nut1414",
+		"id": "179876009615556609"
+	},
+	"service": "jpdb",
+	"description": {
+		"en": "What is jpdb? A powerful Japanese dictionary, all-in-one learning system"
```
Comment (ImRodry) [inline websites/J/jpdb/dist/metadata.json:9]:
```suggestion
		"en": "Powerful Japanese dictionary and an all-in-one learning system for kanji and vocabulary."
```

File: websites/J/jpdb/presence.ts [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+	"/live-action",
+	"/video-game",
+	"/web-novel",
+	"/visual-novel",
+	"/audio",
+];
+presence.on("UpdateData", async () => {
+	const presenceData: PresenceData = {
+		largeImageKey: "jpdb",
+		smallImageText: "jpdb.io",
```
Comment (ImRodry) [inline websites/J/jpdb/presence.ts:21]: this is not needed

File: websites/J/jpdb/presence.ts [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+		startTimestamp: browsingTimestamp,
+	};
+	if (document.location.hostname === "jpdb.io") {
```
Comment (ImRodry) [inline websites/J/jpdb/presence.ts:24]: this is redundant

File: websites/J/jpdb/presence.ts [hunk 4]
```diff
@@ -0,0 +1,100 @@
+const presence = new Presence({
+		clientId: "988699263775154176",
+	}),
+	browsingTimestamp = Math.floor(Date.now() / 1000);
+let dueCountElem: Element, deckTitle: string, query: string, meaning: string;
```
Comment (ImRodry) [inline websites/J/jpdb/presence.ts:5]: none of these should be declared globally as you are not gaining anything from doing that. Please remove them

File: websites/J/jpdb/presence.ts [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+		largeImageKey: "jpdb",
+		smallImageText: "jpdb.io",
+		startTimestamp: browsingTimestamp,
+	};
+	if (document.location.hostname === "jpdb.io") {
+		if (document.location.pathname.includes("/learn")) {
+			presenceData.details = "Viewing learn page";
+			dueCountElem =
+				document.querySelector('a[href="/learn"]').firstElementChild;
+			if (dueCountElem.getAttribute("style") === "color: green;")
```
Comment (ImRodry) [inline websites/J/jpdb/presence.ts:29]: this seems unstable, is there no element or text you can check for instead?

File: websites/J/jpdb/presence.ts [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+				presenceData.state = `New : ${dueCountElem.textContent}\n items`;
+			else presenceData.state = `Due : ${dueCountElem.textContent}\n items`;
```
Comment (ImRodry) [inline websites/J/jpdb/presence.ts:31]:
```suggestion
				presenceData.state = `New: ${dueCountElem.textContent} items`;
			else presenceData.state = `Due: ${dueCountElem.textContent} items`;
```

File: websites/J/jpdb/presence.ts [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+					presenceData.details = "Viewing a kanji:";
+					query = new URLSearchParams(document.location.search).get("q");
+					meaning = decodeURI(
+						document.querySelector(
+							"div.result.kanji > div.vbox.gap > div.hbox > div.vbox.gap > div > div"
+						).textContent
+					);
+					presenceData.state = `${query} - ${meaning}`;
+				}
+				if (document.querySelector("div.result.vocabulary")) {
```
Comment (ImRodry) [inline websites/J/jpdb/presence.ts:57]: surely this should be an else if

**Review by ImRodry — CHANGES_REQUESTED** [review]
Also please add support for all pages on the website. To name a few that are missing: login and the following ones:
image
image

File: websites/J/jpdb/presence.ts (hunk 6 above)
Comment (nut1414) [reply websites/J/jpdb/presence.ts:31]: i thought i already reverted the `\n` oopsies :p

File: websites/J/jpdb/presence.ts (hunk 7 above)
Comment (nut1414) [reply websites/J/jpdb/presence.ts:57]: yes indeed.

File: websites/J/jpdb/presence.ts (hunk 5 above)
Comment (nut1414) [reply websites/J/jpdb/presence.ts:29]: while there are text showing on `/learn` endpoint, I think it's more consistent to use navbar text considering that `/review` endpoint doesn't have any other text than the link on the navbar?
image

File: websites/J/jpdb/presence.ts (hunk 4 above)
Comment (nut1414) [reply websites/J/jpdb/presence.ts:5]: should `decklist` be in the check instead too?
```ts
if ([
      "/novel/",
      "/vocabulary-list/",
      "/textbook/",
      "/anime/",
      "/youtube-video/",
      "/live-action/",
      "/video-game/",
      "/web-novel/",
      "/visual-novel/",
      "/audio/",
  ].some(deck => document.location.pathname.includes(deck))
}
```

**Comment by nut1414** [comment]
this should hopefully addressed most of the changes
<details>
<summary> Proof showing the creation/modification is working as expected </summary>

ndl
privacy
tos
vdl
wnl
about
adl
changelog
contact
faq
ldl
login

</details>

**Review by nut1414 — COMMENTED** [review]
now include page that can only accessed by using the following links
https://jpdb.io/analyze-text
https://jpdb.io/kanken-kanji
https://jpdb.io/kanji-by-frequency

File: websites/J/jpdb/presence.ts (hunk 4 above)
Comment (ImRodry) [reply websites/J/jpdb/presence.ts]: you can keep that one as a global constant. I'm only saying you don't need the variables

**Comment by nut1414** [comment]
<details>
<summary> secret pages </summary>

kanken
freqlist
analy


</details>

<details>
<summary> deck importing pages </summary>

addfromtext
addfromtextconfirm
createdeck
createdecktext
createdecktopvocab
importanki
importankideck
importankioverview
importsatori
importshirabe

</details>

File: websites/J/jpdb/presence.ts [hunk 8]
```diff
@@ -0,0 +1,172 @@
...
+		if (dueCountElem.getAttribute("style") === "color: green;")
+			presenceData.state = `New: ${dueCountElem.textContent} items`;
+		else presenceData.state = `Due: ${dueCountElem.textContent} items`;
+	} else if (document.location.pathname.includes("/review")) {
+		presenceData.details = "Reviewing cards";
+		const dueCountElem =
+			document.querySelector('a[href="/learn"]').firstElementChild;
+		if (dueCountElem.getAttribute("style") === "color: green;")
+			presenceData.state = `New: ${dueCountElem.textContent}\n items`;
+		else presenceData.state = `Due : ${dueCountElem.textContent}\n items`;
```
Comment (EncryptedDev) [inline websites/J/jpdb/presence.ts:36]: What's the new line bits here for? (`\n`)

File: websites/J/jpdb/presence.ts [hunk 9]
```diff
@@ -0,0 +1,172 @@
...
+	else if (document.location.pathname.includes("/novel-difficulty-list"))
+		presenceData.details = "Viewing novel difficulty list";
+	else if (document.location.pathname.includes("/web-novel-difficulty-list"))
+		presenceData.details = "Viewing web novel difficulty list";
+	else if (document.location.pathname.includes("/kanji-by-frequency"))
+		presenceData.details = "Viewing kanji by frequency list";
+	else if (document.location.pathname.includes("/kanken-kanji"))
+		presenceData.details = "Viewing Kanken kanji list";
+	else if (document.location.pathname.includes("/analyze-text"))
+		presenceData.details = "Analyzing japanese text";
```
Comment (EncryptedDev) [inline websites/J/jpdb/presence.ts:156]:
```suggestion
		presenceData.details = "Analyzing Japanese text";
```

**Review by EncryptedDev — CHANGES_REQUESTED** [review]

**Review by EncryptedDev — APPROVED** [review]
LGTM

**Review by Bas950 — APPROVED** [review]
LGTM

**Pull request merged**